### PR TITLE
ad layoutCallback support render-start

### PR DESCRIPTION
**Pull request by zhouyx**
separate #4284 to multiple PRs. This is the one that integrate render-start promise to layoutCallback promise.

#4586

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 1]
```diff
@@ -63,7 +64,7 @@ export class AmpAdApiHandler {
 
   /**
    * Sets up listeners and iframe state for iframe containing ad creative.
-   * @param {!Element} iframe
+
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: woops, will add this back

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 2]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
       // creative as it will not expect having to send the render-start message.
       this.iframe_.style.visibility = 'hidden';
     }
-    listenForOnce(this.iframe_, 'render-start', () => {
-      if (!this.iframe_) {
-        return;
+
+    const promise = new Promise(resolve => {
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: This would be handy as a helper in the `iframe-helper` module.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 3]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
...
+      listenForOnce(this.iframe_, 'render-start', () => {
+        resolve();
+      }, this.is3p_);
+    });
+
+    this.renderStartPromise_ = timerFor(this.baseInstance_.win).timeoutPromise(
+        2000, promise,
+        'fail to receive render-start event from ad server, timeout');
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: `from ad server` is inaccurate.

you can just say `render-start event timed out`

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 4]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
...
-        this.baseInstance_.renderStartResolve_();
-        this.baseInstance_.renderStartResolve_ = null;
+    }).catch(() => {
+      //TODO: report performance
+      if (this.iframe_) {
+        this.iframe_.style.visibility = '';
       }
-      this.iframe_.style.visibility = '';
-    }, this.is3p_);
+    });
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: If we always do the same thing in catch and then, the code can be simplified:
```
.catch(() => {})
.then(() => {
  ...
});
```

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 5]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
...
+
     this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     });
     this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_);
+    return loadPromise(this.iframe_).then(() => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]:
```
return Promise.all([loadPromise(), renderStartPromise]);
```

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 6]
```diff
@@ -35,39 +36,87 @@ describe('amp-ad-api-handler', () => {
...
+          apiHandler.iframe_.src = iframeSrc;
+        };
+        return apiHandler.renderStartPromise_.then(() => {
+          return startUpPromise.then(() => {
+            expect(apiHandler.iframe_.style.visibility).to.equal('');
+          });
+        }, () => {
+          throw new Error('should never happen');
+        });
+      });
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: As we discussed, please wait for #4583 and restructure the tests.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 7]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
       // creative as it will not expect having to send the render-start message.
       this.iframe_.style.visibility = 'hidden';
     }
-    listenForOnce(this.iframe_, 'render-start', () => {
-      if (!this.iframe_) {
-        return;
+
+    const promise = new Promise(resolve => {
+      listenForOnce(this.iframe_, 'render-start', () => {
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Just `listenForOnce(this.iframe_, 'render-start', resolve)`

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 8]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
...
+        resolve();
+      }, this.is3p_);
+    });
+
+    this.renderStartPromise_ = timerFor(this.baseInstance_.win).timeoutPromise(
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: This is not the right time to set up timeout. We need timeout from the time when the iframe was loaded, not from the time we start loading it.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 9]
```diff
@@ -121,21 +122,36 @@ export class AmpAdApiHandler {
...
+      if (this.iframe_) {
+        this.iframe_.style.visibility = '';
       }
-      if (this.baseInstance_.renderStartResolve_) {
-        this.baseInstance_.renderStartResolve_();
-        this.baseInstance_.renderStartResolve_ = null;
+    }).catch(() => {
+      //TODO: report performance
+      if (this.iframe_) {
+        this.iframe_.style.visibility = '';
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Either `dev.error` the error here or rethrow it.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 10]
```diff
@@ -35,39 +36,87 @@ describe('amp-ad-api-handler', () => {
     const adElement = document.createElement('amp-ad');
     const adImpl = new BaseElement(adElement);
     apiHandler = new AmpAdApiHandler(adImpl, adImpl.element);
-    return createIframePromise().then(c => {
+    return createIframePromise(null, () => {
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: It looks like `null, () => {}` are not needed here - a leftover from our attempts to fix tests

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 11]
```diff
@@ -24,6 +24,7 @@ describe('amp-ad-api-handler', () => {
   let container;
   let iframe;
   let apiHandler;
+  let startUpPromise;
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: This should move to the `describe` that actually creates it.

File: test/functional/test-element-stub.js [hunk 12]
```diff
@@ -15,30 +15,35 @@
...
 
 import {ElementStub, resetLoadingCheckForTests} from '../../src/element-stub';
-import {
-    createIframePromise,
-    doNotLoadExternalResourcesInTest,
-} from '../../testing/iframe';
+import {createIframePromise} from '../../testing/iframe';
 import '../../extensions/amp-ad/0.1/amp-ad';
 import '../../extensions/amp-analytics/0.1/amp-analytics';
+import {installPerformanceService,} from
```
Comment (dvoytenko) [inline test/functional/test-element-stub.js]: Why is this test modified here?

Comment (zhouyx) [inline test/functional/test-element-stub.js]: nah, leftover from the previous PR (which I add performance report).

remove `doNotLoadExternalResourcesInTest` because I want the integration.js to send `render-start` to resolve `layoutCallback`

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 5 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Since we want to create `timeoutPromise` after this.iframe_ is loaded, I can't do return promise.all here.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 8 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Ah! You're right! changed

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 9 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: I think `timeoutPromise` function will throw the error for us. Why would we rethrow it again?

Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: It does. And when you `catch` it, you actually kill the error. So we need to make sure it gets reported one way or another.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 13]
```diff
@@ -121,21 +130,32 @@ export class AmpAdApiHandler {
...
+        //TODO: report performance
+        if (this.iframe_) {
+          this.iframe_.style.visibility = '';
+        }
+      }).catch(error => {
+        //TODO: report performance
+        if (this.iframe_) {
+          this.iframe_.style.visibility = '';
+        }
+        dev().error(error);
```
Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: `error` call requires you to pass tag. You can always simply rethrow the error as `throw error`;

Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Oh, I see. You want it to always to succeed? In this case, `dev().error(TAG, error)` is the right way.

Comment (dvoytenko) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: You may also want to rewrite it this way:

```
this.renderStartTimeoutPromise_.catch(error = > {
  dev().error(TAG, error);
}).then(() => {
  if (this.iframe_) {
    this.iframe_.style.visibility = '';
  }
});
```

Further, it's most likely incorrect to sub-chain the promise here and then return `return this. renderStartTimeoutPromise_;`  - basically what happens is that you will return loading promise before you completed all loading actions.

And finally, do you actually need an instance var? Why not just do:
```
return timer
    .timeoutPromise(200, this.renderStartPromise_, ...)
    .catch(error => { /* report error */ })
    .then(() => { /* complete loading/init re:visibility */});
```

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 14]
```diff
@@ -251,10 +243,12 @@ export class AmpAd3PImpl extends AMP.BaseElement {
         this.iframe_ = getIframe(this.element.ownerDocument.defaultView,
             this.element);
         this.apiHandler_ = new AmpAdApiHandler(
-          this, this.element, this.boundNoContentHandler_);
+            this, this.element, this.boundNoContentHandler_);
         return this.apiHandler_.startUp(this.iframe_, true);
       });
     }
+    dev().error('AMP_AD',
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Hm?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 15]
```diff
@@ -121,21 +130,32 @@ export class AmpAdApiHandler {
...
+    this.renderStartPromise_ =
+        listenForOncePromise(this.iframe_, 'render-start', this.is3p_);
+
     this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     });
     this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_);
+    return loadPromise(this.iframe_).then(() => {
+      this.renderStartTimeoutPromise_ = timerFor(this.baseInstance_.win)
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Does this need to be a property?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 16]
```diff
@@ -121,21 +130,32 @@ export class AmpAdApiHandler {
...
+          .timeoutPromise(200, this.renderStartPromise_,
+          'render-start-event timed out');
+      this.renderStartTimeoutPromise_.then(() => {
+        //TODO: report performance
+        if (this.iframe_) {
+          this.iframe_.style.visibility = '';
+        }
+      }).catch(error => {
+        //TODO: report performance
+        if (this.iframe_) {
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: How would we have an iframe if there was an error?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 14 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: I thought ad layoutCallback should always create a new iframe. what are cases where we schedule layout for `amp-ad` with an existing iframe?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 16 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: It's the error for `renderStartTimeoutPromise_`, not for `loadPromise`

Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: There're still be a `this.iframe_` in that case.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js (hunk 13 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: No, I want it to fail when timeout. Changed the code

**Comment by zhouyx** [comment]
PTAL

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 14 above)
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: This seems like the wrong place to assert this, then.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 17]
```diff
@@ -59,6 +61,12 @@ export class AmpAdApiHandler {
 
     /** @private @const */
     this.viewer_ = viewerFor(this.baseInstance_.win);
+
+    /** @private {!Promise|null} */
+    this.renderStartPromise_ = null;
+
+    /** @private {!Promise|null} */
+    this.renderStartTimeoutPromise_ = null;
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Unused.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 18]
```diff
@@ -121,21 +129,31 @@ export class AmpAdApiHandler {
...
+    return loadPromise(this.iframe_).then(() => {
+      return timerFor(this.baseInstance_.win)
+          .timeoutPromise(200, this.renderStartPromise_,
+          'render-start-event timed out')
+          .catch(error => {
+            //TODO: report performance
+            if (this.iframe_) {
+              this.iframe_.style.visibility = '';
+            }
+            throw error;
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Why are we erroring here?

Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: never receive render-start, I think we should fail the startUp promise and throw the error to `layoutCallback`

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js (hunk 14 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Could you be more clear about where I should put it?  Here's what I do now.
``` 
layoutCallback() {
  if (!this.iframe_) {
    //create iframe 
    return loadPromise(); 
  }
  dev().error();
  return loadPromise();
}
```

**Comment by zhouyx** [comment]
Talked to @lannka. Timeout will get refactored anyway by introducing the race condition of `render-start` and `no-content`. So I decided not to add timeout in this PR.

File: test/fixtures/served/iframe.html [hunk 19]
```diff
@@ -40,6 +40,13 @@
             sentinel: sentinel,
             type: 'send-embed-state',
           }, '*');
+    } else if (event.data.type == 'renderStart') {
```
Comment (lannka) [inline test/fixtures/served/iframe.html]: is this change still needed?

File: test/functional/test-ad-cid.js [hunk 20]
```diff
@@ -52,16 +52,16 @@ function tests(name) {
 
       afterEach(() => {
         sandbox.restore();
-        delete clientIdScope['with_cid'];
+        delete clientIdScope['_ping_'];
```
Comment (lannka) [inline test/functional/test-ad-cid.js]: is this related to this change?

File: test/fixtures/served/iframe.html (hunk 19 above)
Comment (zhouyx) [inline test/fixtures/served/iframe.html]: no needed , will delete

File: src/iframe-helper.js [hunk 21]
```diff
@@ -311,6 +311,20 @@ export function listenForOnce(iframe, typeOfMessage, callback, opt_is3P) {
...
 
 /**
+ * Return a promise that resolve when listenForOnce function has listened to the message
+ * @param {!Element} iframe
+ * @param {string} typeOfMessage
+ * @param {boolean=} opt_is3P
+ * @return {!Promise}
+ */
+export function listenForOncePromise(iframe, typeOfMessage, opt_is3P) {
+  const promise = new Promise(resolve => {
```
Comment (lannka) [inline src/iframe-helper.js]: nit: `return new Promise(...`

File: src/iframe-helper.js [hunk 22]
```diff
@@ -311,6 +311,20 @@ export function listenForOnce(iframe, typeOfMessage, callback, opt_is3P) {
 }
 
 /**
+ * Return a promise that resolve when listenForOnce function has listened to the message
```
Comment (lannka) [inline src/iframe-helper.js]: `Returns a promise that resolves when the given message has been observed for the first time.`

File: test/functional/test-ad-cid.js (hunk 20 above)
Comment (zhouyx) [inline test/functional/test-ad-cid.js]: ad `layoutCallback` needs `render-start` to resolve.  'with-cid' is an ad type that we don't recognize, and we won't generate `render-start` for it.  Thus getAd() will never succeed and we will get timeout for this test.  
I changed it to '_ping_' to solve this issue.

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 23]
```diff
@@ -30,37 +29,55 @@ describe('amp-ad-api-handler', () => {
...
     sandbox.restore();
     apiHandler = null;
-    testIndex++;
   });
 
-  it('embed-state API', () => {
-    const embedStateData = {
-      sentinel: 'amp3ptest' + testIndex,
-      type: 'send-embed-state',
+  describe('startUp test', () => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: `iframe that is initialized by startUp()`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 24]
```diff
@@ -30,37 +29,55 @@ describe('amp-ad-api-handler', () => {
...
+          iframe = newIframe;
+        });
+    });
 
-    const embedStateDataReply = 'amp-' + JSON.stringify({
-      inViewport: false,
-      pageHidden: false,
-      type: 'embed-state',
-      sentinel: 'amp3ptest' + testIndex,
+    it('embed-state API', () => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: `should be able to use `embed-state` API`

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 25]
```diff
@@ -30,37 +29,55 @@ describe('amp-ad-api-handler', () => {
...
+      const embedStateData = {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: nit: inline the variable

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 26]
```diff
@@ -30,37 +29,55 @@ describe('amp-ad-api-handler', () => {
...
+        sentinel: 'amp3ptest' + testIndex,
+        type: 'send-embed-state',
+      };
+
+      const embedStateDataReply = 'amp-' + JSON.stringify({
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: nit: inline the variable

File: extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js [hunk 27]
```diff
@@ -30,37 +29,55 @@ describe('amp-ad-api-handler', () => {
...
+        sentinel: 'amp3ptest' + testIndex,
+      });
+
+      iframe.postMessageToParent(embedStateData);
+      return iframe.expectMessageFromParent(embedStateDataReply);
     });
 
-    iframe.postMessageToParent(embedStateData);
-    return iframe.expectMessageFromParent(embedStateDataReply);
+    it('render-start API resolve promise', () => {
```
Comment (lannka) [inline extensions/amp-ad/0.1/test/test-amp-ad-api-handler.js]: should resolve startUp() when `render-start` API is called

**Comment by lannka** [comment]
LGTM with some minor comments.
